lib/Toplevel_lexer.mll

```diff
@@ -0,0 +1,67 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                              OCamlFormat                               *)
+(*                                                                        *)
+(*            Copyright (c) Facebook, Inc. and its affiliates.            *)
+(*                                                                        *)
+(*      This source code is licensed under the MIT license found in       *)
+(*      the LICENSE file in the root directory of this source tree.       *)
+(*                                                                        *)
+(**************************************************************************)
+
+{
+open Caml
+
+let newline lexbuf = Lexing.new_line lexbuf
+}
+
+let eol = '\n' | eof
+let ws = ' ' | '\t'
+
+rule token = parse
+ | eof           { [] }
+ | '\n'          { newline lexbuf; `Output "" :: token lexbuf }
+ | "# "          { let loc = Location.curr lexbuf in
+                   let c = phrase [] (Buffer.create 8) lexbuf in
+                   `Command (c, loc) :: token lexbuf }
+ | ([^'#' '\n'] [^'\n']* as str) eol
+                 { newline lexbuf; `Output str :: token lexbuf }
+ | _ as c        { failwith (Printf.sprintf "unexpected character '%c'. Did you forget a space after the '#' at the start of the line?" c) }
+
+and phrase acc buf = parse
+  | ("\n"* as nl) "\n" ("  " | "\t")
+      { newline lexbuf;
+        for _ = 1 to (String.length nl) do
+          newline lexbuf
+        done;
+        let nl = List.init (String.length nl) (fun _ -> "") in
+        phrase (nl @ Buffer.contents buf :: acc) (Buffer.create 8) lexbuf }
+  | eol      { newline lexbuf; List.rev (Buffer.contents buf :: acc) }
+  | ";;" eol { newline lexbuf; List.rev ((Buffer.contents buf ^ ";;") :: acc) }
+  | _ as c   { Buffer.add_char buf c; phrase acc buf lexbuf }
+
+{
+let repl_file lx =
+  let x =
+    try
+      newline lx;
+      token lx
+    with _ -> raise (Syntaxerr.Error (Expecting (Location.curr lx, ";;")))
+  in
+  let open Ocaml_413_extended.Parsetree in
+  List.fold_left (fun acc -> function
+      | `Command (lines, _loc) ->
+          { prepl_phrase=
+              String.concat "\n" lines
+              |> Lexing.from_string |> Parse.toplevel_phrase
+          ; prepl_output= "" }
+          :: acc
+      | `Output line -> (
+          match acc with
+          | [] -> raise (Syntaxerr.Error (Expecting (Location.curr lx, "#")))
```

Comment:
I've pushed a failing test related to this in https://github.com/gpetiot/ocamlformat/pull/7

Reply:
Merged, thanks!

Reply:
It's still failing. In `test/cli/repl_file_errors.t/run.t` it's `empty_line_begin.repl`.

Reply:
I think havint just `Syntax error` printed comes from the location not being properly set in the exception, I improved the situation a bit but it looks like it's not fixed yet

Reply:
@julow I tried a few things, but the locations are invalid anyway, since we have locations relative to the opening of the block instead of the start of the file. `odoc-parser` is attaching a `Loc.span` to the code block but it's not enough to get back a `Lexing.position`.
Since we don't modify invalid blocks it's not really a must-have to print a warning. We can open another PR if there is a way to get a position in the future.

Reply:
That's not ideal but it might be possible to craft a `Lexing.position` value from odoc's location.
Otherwise, reporting the location of the entire comment is more useful than a completely wrong location, I think.

Reply:
I added a warning when the parsing of toplevel block failed